### ISSUE TITLE
terraform provider for netbird REST API

### ISSUE THREAD

**Issue by theblop** [netbirdio/netbird#1208]
**Is your feature request related to a problem? Please describe.**
It would be great to be able to manage netbird server with terraform (a lot of people already use terraform to manage their infrastructure)

**Describe the solution you'd like**
a terraform provider to interact with the netbird REST API

**Comment by mlsmaycon** [comment]
Hello @theblop, thanks for the request. We are looking at a terraform provider initial version for this quarter. We will keep this updated as we progress with it.

**Comment by mkoyan44** [comment]
@mlsmaycon 
Hey, do you have any updates on the topic? Which resources are you going to support?

**Comment by ednxzu** [comment]
Any update on this ? Would be super convenient !

**Comment by boopzz** [comment]
looks like speakeasy make terraform providers for projects so it looks like the start, hopefully theres a private branch just about to be committed! 🙏

**Comment by tafaust** [comment]
Is it safe (and legal) to use that codebase as a foundation @speakeasy-sdks / @TristanSpeakEasy?

**Comment by mlsmaycon** [comment]
We have a community member working on a provider. We are using the Hashicorp OpenAPI Provider Spec Generator for the foundation. 

We should have an initial version of the provider earlier in Q4 this year.

**Comment by simplesagar** [comment]
Hey @tafaust that repository is meant to be an example of our generation capabilities at https://github.com/speakeasy-api and may not be up to date with the latest Netbird API

@mlsmaycon would be happy to help you all generate and maintain a really nice terraform provider. Here are a few we maintain in real time. 
- https://github.com/opalsecurity/terraform-provider-opal
- https://github.com/StyraInc/terraform-provider-styra 
- https://github.com/airbytehq/terraform-provider-airbyte

Here's a [walk through](https://www.speakeasy.com/post/how-to-build-terraform-providers) of our generation as well as Hashicorp's method. Would love to help if we can !

**Comment by marcportabellaclotet-mt** [comment]
Using Speakeasy makes it incredibly simple to create a Terraform provider for Netbird. After some testing, I was able to build a functional provider. However, there are two aspects of the Netbird API that might need addressing:

1. **Route Creation**: Currently, the Netbird API allows defining either a network or a domain when creating a route. However, if an empty network value is provided while specifying a domain, the API returns an error, as only one of the two is allowed. Could this behavior be adjusted to accept an empty network value when a domain is defined? This change would simplify Terraform provider development, though it is possible to implement a workaround.

2. **Setup Key Values**: When setup keys are created, their values are returned in plain text. However, subsequent API responses for these keys provide only obfuscated values. This causes the Terraform provider to retrieve only the obfuscated version, which can also be worked around but adds complexity.

According to Speakeasy's documentation, their service might be free for open-source projects. Would the Netbird team be interested in exploring this option?

(Text reference: [[Speakeasy Pricing](https://www.speakeasy.com/pricing)](https://www.speakeasy.com/pricing))
```
If you're working on a 'free as in speech' OSS project, you can use Speakeasy for free. Just fill out this TypeForm: https://speakeasyapi.typeform.com/oss-program
```

**Comment by ayprof** [comment]
Any news?

**Comment by tafaust** [comment]
@mlsmaycon for reference. Maybe Speakeasy is the way to go here? I guess the terraform provider will not be monetised anyways?

**Comment by sebastienbaillet** [comment]
Any news ?

**Comment by nazarewk** [comment]
We're still trying to organize an official provider, but we don't have an ETA. Personally, I currently have more than enough responsibilities during work hours and cannot dedicate overtime to it.

Meanwhile, here is a non-exhaustive list of community efforts:
- [recent Slack thread](https://netbirdio.slack.com/archives/C05T5K65X7U/p1743086753289469)
- https://registry.terraform.io/providers/DockStudios/netbird/latest
- https://github.com/mojaloop/terraform-provider-netbird

**Comment by praneetloke** [comment]
I have a kind of a related question. I am looking into generating a Pulumi provider from your OpenAPI specs but I couldn't the OpenAPI spec in this repo. Your docs website don't mention anything either.

**Comment by mlsmaycon** [comment]
Hey folks, regarding Terraform we have a provider for some time, and somehow, didn't close this issue: 

https://github.com/netbirdio/terraform-provider-netbird
https://registry.terraform.io/providers/netbirdio/netbird/latest/docs

@praneetloke it uses our REST client: 

https://github.com/netbirdio/netbird/tree/main/shared/management/client/rest

We are also updating it:
https://github.com/netbirdio/netbird/pull/5222

@praneetloke, I will close this one as completed, but we can discuss the Pulumi provider in another issue if you want. That interests us very much.

**Comment by praneetloke** [comment]
@mlsmaycon thank you! I've subscribed to that PR. Out of curiosity, do you use a generator for the OpenAPI spec? I do see couple of issues in the API spec that is in #5222. I'll wait for it to be merged and check again.

**Comment by mlsmaycon** [comment]
@praneetloke the generators are next to the spec: https://github.com/netbirdio/netbird/blob/main/shared/management/http/api/generate.sh

**Comment by praneetloke** [comment]
Got it. Thank you. I was specifically interested in how the OpenAPI spec, i.e. [`openapi.yml`](https://github.com/netbirdio/netbird/blob/main/shared/management/http/api/openapi.yml), was maintained. Do you use any tools to generate/validate that spec doc?

**Comment by mlsmaycon** [comment]
Not really. We usually create the endpoint spec manually and generate the API types out of it.

**Comment by praneetloke** [comment]
I'll open a new issue with some observations. Thank you for your responses, @mlsmaycon!
